protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -81,148 +51,240 @@ where
 {
     type ConnectionHandler = Handler;
 
-    type ToSwarm = Event;
+    type ToSwarm = ();
 
     fn handle_established_inbound_connection(
         &mut self,
-        _connection_id: ConnectionId,
-        peer: PeerId,
+        connection_id: ConnectionId,
+        _peer: PeerId,
         _local_addr: &Multiaddr,
-        _remote_addr: &Multiaddr,
-    ) -> Result<Self::ConnectionHandler, ConnectionDenied> {
-        Ok(Handler::new(self.accepted_nonce.clone(), peer))
+        remote_addr: &Multiaddr,
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        if addr_is_local(remote_addr) {
+            self.local_peers.insert(connection_id);
+        }
+        Ok(new_handler())
     }
 
     fn handle_established_outbound_connection(
         &mut self,
-        _connection_id: ConnectionId,
-        peer: PeerId,
-        _addr: &Multiaddr,
+        connection_id: ConnectionId,
+        _peer: PeerId,
+        addr: &Multiaddr,
         _role_override: Endpoint,
         _port_use: PortUse,
-    ) -> Result<Self::ConnectionHandler, ConnectionDenied> {
-        Ok(Handler::new(self.accepted_nonce.clone(), peer))
+    ) -> Result<<Self as NetworkBehaviour>::ConnectionHandler, ConnectionDenied> {
+        if addr_is_local(addr) {
+            self.local_peers.insert(connection_id);
+        }
+        Ok(new_handler())
     }
 
-    fn on_swarm_event(&mut self, _event: FromSwarm) {
-        todo!()
+    fn on_swarm_event(&mut self, event: FromSwarm) {
+        match event {
+            FromSwarm::NewExternalAddrCandidate(NewExternalAddrCandidate { addr }) => {
+                for _ in 0..self.config.test_server_count {
+                    let nonce = self.rng.gen();
+                    self.pending_requests.push_back(DialRequest {
+                        addrs: vec![addr.clone()],
+                        nonce,
+                    });
+                    self.pending_nonces.insert(nonce);
+                }
```

Comment:
We'll need some kind of throttling here, otherwise a malicious peer can trigger load on other nodes by sending us a different observed address via identify on every identify request.

My suggestion would be that we:

- Accumulate all candidates in the behaviour state and rank them by occurrence. More frequently reported candidates are more likely to be external addresses and should thus be reported first.
- Filter out candidates that we already confirmed as external addresses.
- On a timer, take the first N candidates and send dial-requests to the currently connected autonat servers.

For now, we can simply assume that the connection will be kept-alive. In our tests, we can achieve this by setting an idle connection timeout that is longer than the timer. In a real application, connections come and go all the time or stay alive for something like gossipsub or an active reservation with a relay. We can later also think about optimisations to e.g. check if we still have pending candidates even though the timer fired already multiple times and we just connected to a new autonat server.

Like I said above, it is fine to assume we have a connection from _somewhere_ for now but it is important that we implement this throttling :)